duration,builtins: fix DST boundary handling

otan opened the pull request:
See individual commits for details.

Resolves #64772

otan committed:
duration: fix Add on DST boundaries

time.Date cannot differentiate times if they have two possible values
for a given time in a given location. For example, 2020-12-25 can have
03:00+02 and 03:00+03 in Europe/Bucharest and time.Date can choose
either way (I believe it chooses whatever is later).

This breaks timestamptz / interval math if we use AddDate on that
boundary. This has been fixed by only using AddDate if day / months > 0,
in which case the library will do the correct thing.

Release note (bug fix): Fix a bug where interval math on a TIMESTAMPTZ
on a DST boundary would incorrectly add or subtract an extra hour.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/64845)
<!-- Reviewable:end -->

RaduBerinde reviewed (commented):
Nice fixes, thanks for taking this on!!

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64845#-:-M_2pw3l7jzzLRitZFBn:b-9wynpi)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/logictest/testdata/logic_test/timestamp, line 498 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64845#-M_2pdkh1soCnkowuULf:-M_2pdkh1soCnkowuULg:b-oksjrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1b05e25e32b1f9a6f5d1cad2fcf2b03f4248dd0/pkg/sql/logictest/testdata/logic_test/timestamp#L498)):*
> ```
> query T
> SET TIME ZONE 'Europe/Bucharest';
> SELECT t FROM ( VALUES
> ```

[nit] extra space before VALUES


<!-- Sent from Reviewable.io -->

otan committed:
builtins: fix date_trunc on DST boundaries

Release note (bug fix): Fix a bug where date_trunc on a time on a
DST boundary could switch timezones and produce the incorrect result.

otan commented:
tftr!

bors r=raduberinde
